plugins/PrivacyManager/Model/DataSubjects.php

```diff
@@ -114,6 +119,51 @@ public function deleteDataSubjects($visits)
         return $results;
     }
 
+    private function invalidateArchives($visits)
+    {
+        $datesToInvalidateByIdSite = $this->getDatesToInvalidate($visits);
+
+        $invalidator = StaticContainer::get('Piwik\Archive\ArchiveInvalidator');
+
+        foreach ($datesToInvalidateByIdSite as $idSite => $visitDates) {
+            foreach ($visitDates as $dateStr) {
+                $visitDate = Date::factory($dateStr);
+                $invalidator->rememberToInvalidateArchivedReportsLater($idSite, $visitDate);
+            }
+        }
+    }
+
+    private function getDatesToInvalidate($visits)
+    {
+        $idVisitsByIdSites = array();
+        foreach ($visits as $visit) {
+            $idSite = (int)$visit['idsite'];
+            if (!isset($idVisitsByIdSites[$idSite])) {
+                $idVisitsByIdSites[$idSite] = array();
+            }
+            $idVisitsByIdSites[$idSite][] = (int)$visit['idvisit'];
+        }
+
+        $datesToInvalidate = array();
+        foreach ($idVisitsByIdSites as $idSite => $idVisits) {
+            $website = Cache::getCacheWebsiteAttributes($idSite);
+            $timezone = $website['timezone'];
+
+            $sql = 'SELECT visit_last_action_time FROM '
+                . Common::prefixTable('log_visit') . ' WHERE idsite = ' . $idSite
```

Comment:
@katebutler be good to cast `idSite` to `(int)` or use bind parameters just to be double safe. Can basically never be an issue here since we have checked above it is an idSite but better to be safe :)

Reply:
I'll delete this comment, it's already an integer :)